contracts/retirement/KlimaRetirementAggregator.sol

```diff
@@ -0,0 +1,388 @@
+
+import "../interfaces/IUniswapV2Router02.sol";
+import "../interfaces/IwsKLIMA.sol";
+import "../interfaces/IRetireBridgeCommon.sol";
+import "../interfaces/IRetireMossCarbon.sol";
+import "../interfaces/IRetireToucanCarbon.sol";
+
+/**
+ * @title KlimaRetirementAggregator
+ * @author KlimaDAO
+ *
+ * @notice This is the master aggregator contract for the Klima retirement utility.
+ *
+ * This allows a user to provide a source token and an approved carbon pool token to retire.
+ * If the source is different than the pool, it will attempt to swap to that pool then retire.
+ */
+contract KlimaRetirementAggregator is
+    Initializable,
+    ContextUpgradeable,
+    OwnableUpgradeable
+{
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    function initialize() public initializer {
+        __Ownable_init();
+        __Context_init();
+    }
+
+    /** === State Variables and Mappings === */
+    address public KLIMA;
+    address public sKLIMA;
+    address public wsKLIMA;
+    address public USDC;
+    address public staking;
+    address public stakingHelper;
+    address public treasury;
+    address public klimaRetirementStorage;
+
+    mapping(address => bool) public isPoolToken;
+    mapping(address => uint256) public poolBridge;
+    mapping(uint256 => address) public bridgeHelper;
+
+    /** === Event Setup === */
+    event AddressUpdated(
+        uint256 addressIndex,
+        address indexed oldAddress,
+        address indexed newAddress
+    );
+    event PoolAdded(address poolToken, uint256 bridge);
+    event PoolRemoved(address poolToken);
+    event BridgeHelperUpdated(uint256 bridgeID, address helper);
+
+    /**
+     * @notice This function will retire a carbon pool token that is held
+     * in the caller's wallet. Depending on the pool provided the appropriate
+     * retirement helper will be used as defined in the bridgeHelper mapping.
+     * If a token other than the pool is provided then the helper will attempt
+     * to swap to the appropriate pool and then retire.
+     *
+     * @param _sourceToken The contract address of the token being supplied.
+     * @param _poolToken The contract address of the pool token being retired.
+     * @param _amount The amount being supplied. Expressed in either the total
+     *          carbon to offset or the total source to spend. See _amountInCarbon.
+     * @param _amountInCarbon Bool indicating if _amount is in carbon or source.
+     * @param _beneficiaryAddress Address of the beneficiary of the retirement.
+     * @param _beneficiaryString String representing the beneficiary. A name perhaps.
+     * @param _retirementMessage Specific message relating to this retirement event.
+     */
+    function retireCarbon(
+        address _sourceToken,
+        address _poolToken,
+        uint256 _amount,
+        bool _amountInCarbon,
+        address _beneficiaryAddress,
+        string memory _beneficiaryString,
```

Comment:
Could have used `calldata` here and everywhere else where these strings are originating from a transaction and never mutated.